core/kotlinx-coroutines-core/src/internal/LockFreeTaskQueue.kt

```diff
@@ -198,44 +245,64 @@ internal class LockFreeMPSCQueueCore<E : Any>(private val capacity: Int) {
         return next
     }
 
+    // Used for validation in tests only
+    fun <R> map(transform: (E) -> R): List<R> {
+        val res = ArrayList<R>(array.length())
+        _state.value.withState { head, tail ->
+            var index = head
+            while (index and mask != tail and mask) {
+                // replace nulls with placeholders on copy
+                val element = array[index and mask]
+                @Suppress("UNCHECKED_CAST")
+                if (element != null && element !is Placeholder) res.add(transform(element as E))
+                index++
+            }
+        }
+        return res
+    }
+
+
     // Instance of this class is placed into array when we have to copy array, but addLast is in progress --
     // it had already reserved a slot in the array (with null) and have not yet put its value there.
     // Placeholder keeps the actual index (not masked) to distinguish placeholders on different wraparounds of array
-    private class Placeholder(@JvmField val index: Int)
+    // Internal because of inlining
```

Comment:
it can be safely made private again

Reply:
Nope. Does not work for me:
```
kotlinx.coroutines.scheduling.BlockingCoroutineDispatcherRaceStressTest > testAddPollRace STANDARD_ERROR
    Exception in thread "DefaultDispatcher-worker-1" java.lang.IllegalAccessError: tried to access class kotlinx.coroutines.internal.LockFreeTaskQueueCore$Placeholder from class kotlinx.coroutines.scheduling.GlobalQueue
        at kotlinx.coroutines.scheduling.GlobalQueue.removeFirstBlockingModeOrNull(Tasks.kt:147)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.blockingQuiescence(CoroutineScheduler.kt:911)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.blockingWorkerIdle(CoroutineScheduler.kt:828)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:730)
    Exception in thread "DefaultDispatcher-worker-3" java.lang.IllegalAccessError: kotlinx/coroutines/internal/LockFreeTaskQueueCore$Placeholder
        at kotlinx.coroutines.scheduling.GlobalQueue.removeFirstBlockingModeOrNull(Tasks.kt:147)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.blockingQuiescence(CoroutineScheduler.kt:911)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.blockingWorkerIdle(CoroutineScheduler.kt:828)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:730)
```

todo: need to file a bug to JVM BE.